File: toggle_led_python/toggle_led_python/server_led_panel.py
import rclpy
from rclpy.node import Node
from led_interface.msg import LedTrigger
from led_interface.srv import LedState

class LedServer(Node):
    def __init__(self):
        super().__init__("led_server")
        self.msg = LedTrigger()
        self.declare_parameter("led_state", [False, False, True])
        self.msg.leds = self.get_parameter("led_state").value
        self.server_ = self.create_service(LedState, "set_led", self.led_callback)
        self.publisher_ = self.create_publisher(LedTrigger, "led_panel_state", 10)
        self.hugo = ["full", "empty"]
        self.timer_ = self.create_timer(4, self.pub_on_led)
        self.get_logger().info("Server Led has been initialized")

    def pub_on_led(self):
        self.publisher_.publish(self.msg)

    def led_callback(self, request, response):

        led_number = request.led_number
        state = request.state

        self.get_logger().info(f"State: {state} Led number tete: {led_number}")
        try:
            if state in self.hugo:
                if state == "full":
                    self.msg.leds[led_number-1] = False
                elif state == "empty":
                    self.msg.leds[led_number-1] = True

                response.success = True
                self.pub_on_led()

            else:
                response.success = False

        except Exception as ex:
            self.get_logger().error("Something wrong is not correct %r" %(ex,))
            response.success = False

        return response


def main(args = None):
    rclpy.init(args=args)
    node = LedServer()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__=="__main__":
    main()

File: toggle_led_cpp/src/server_led_panel.cpp
#include "rclcpp/rclcpp.hpp"
#include "led_interface/msg/led_trigger.hpp"
#include "led_interface/srv/led_state.hpp"

class SetLedServer: public rclcpp::Node{

    private:

        rclcpp::Publisher<led_interface::msg::LedTrigger>::SharedPtr publisher_;
        rclcpp::Service<led_interface::srv::LedState>::SharedPtr server_;
        rclcpp::TimerBase::SharedPtr timer_;
        led_interface::msg::LedTrigger msg;

        void call_server(led_interface::srv::LedState::Request::SharedPtr request, 
                    led_interface::srv::LedState::Response::SharedPtr response){

            int led_number = request->led_number;
            std::string state = request->state;
            int size = msg.leds.size();

            try{
                if (led_number <= size && led_number > 0){
                    if (state == "full"||state == "empty"){
                        if (state == "empty"){
                            msg.leds[led_number-1] = true;
                        }
                        else if (state == "full"){
                            msg.leds[led_number-1] = false;
                        }

                        RCLCPP_INFO(this->get_logger(), "Led %d modificated", led_number);
                        response->success = true;
                        led_panel_publish();
                    }

                    else{
                        response->success = false;
                        RCLCPP_ERROR(this->get_logger(), "State argument invalid value. It must be empty or full");
                    }
                }
                else {   

                    RCLCPP_ERROR(this->get_logger(), "ERROR: led_number out of the range");
                    response->success = false;
                }
            }

            catch(std::exception &ex){
                RCLCPP_ERROR(this->get_logger(), "Something wrong is not correct");
                response->success = false;
            }
        }

        void led_panel_publish(){
            publisher_->publish(msg);
        }

    public:

        SetLedServer(): Node("set_led_server"){

            server_ = this->create_service<led_interface::srv::LedState>("set_led", 
                                            std::bind(&SetLedServer::call_server, this,
                                            std::placeholders::_1, std::placeholders::_2));

            publisher_ = this->create_publisher<led_interface::msg::LedTrigger>("led_panel_trigger", 10);
            msg.leds = {false, false, false};
            timer_ = this->create_wall_timer(std::chrono::seconds(4), std::bind(&SetLedServer::led_panel_publish, this));
            RCLCPP_INFO(this->get_logger(), "Led Panel Server has been initialized");
        }
};

int main(int argc, char **argv){

    rclcpp::init(argc, argv);
    auto node = std::make_shared<SetLedServer>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;

}
